File: carbondesign/tags/time_picker.py
"""
Time Picker
===========

See: https://the-carbon-components.netlify.app/?nav=time-picker

""" # pylint:disable=line-too-long
# pylint:disable=too-many-lines
from itertools import chain
#-
from .base import FormNode

class BaseSelectFormNode(FormNode):
    """Base form for select form
    """
    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        required = self.bound_field.field.required and\
                self.bound_field.form.use_required_attribute
        if required:
            values['props'].append(('required', True))

        values['label_class'].append('bx--visually-hidden')
        values['props'].append(('name', self.bound_field.name))


    def render_tmpl_icon_error(self, values, context):
        """Dynamically render a part of the component's template.
        """
        if self.eval(self.kwargs.get('disabled'), context):
            return ''
        return """
<svg focusable="false" preserveAspectRatio="xMidYMid meet"
    xmlns="http://www.w3.org/2000/svg" fill="currentColor"
    class="bx--select__invalid-icon" width="16" height="16"
    viewBox="0 0 16 16" aria-hidden="true">
  <path d="M8,1C4.2,1,1,4.2,1,8s3.2,7,7,7s7-3.1,7-7S11.9,1,8,1z M7.5,4h1v5h-1C7.5,9,7.5,4,7.5,4z M8,12.2	c-0.4,0-0.8-0.4-0.8-0.8s0.3-0.8,0.8-0.8c0.4,0,0.8,0.4,0.8,0.8S8.4,12.2,8,12.2z"></path>
  <path d="M7.5,4h1v5h-1C7.5,9,7.5,4,7.5,4z M8,12.2c-0.4,0-0.8-0.4-0.8-0.8s0.3-0.8,0.8-0.8	c0.4,0,0.8,0.4,0.8,0.8S8.4,12.2,8,12.2z" data-icon-path="inner-path" opacity="0"></path>
</svg>
"""


    def render_tmpl_items(self, values, context):
        """Dynamically render a part of the component's template.
        """
        item_tmpl = """
<option class="bx--select-option" value="{value}" {props}>{child}</option>
"""
        items = []
        for _, value, label in self.choices():
            options = {'value': value, 'child': label}
            props = []
            if self.bound_value and value in self.bound_value:
                props.append('selected')
            options['props'] = ' '.join(props)
            items.append(self.format(item_tmpl, options))

        return '\n'.join(items)


class AmPmPicker(BaseSelectFormNode):
    """AM/PM Picker component

    To be used within TimePicker
    """
    def render_default(self, values, context):
        """Output html of the component.
        """
        if self.bound_field.errors:
            template = """
<div class="bx--time-picker__select bx--select bx--select--invalid {class}"
    data-invalid>
  {tmpl_label}
  <select id="{id}" class="bx--select-input" {props}>
    {tmpl_items}
  </select>
  <svg focusable="false" preserveAspectRatio="xMidYMid meet"
      xmlns="http://www.w3.org/2000/svg" fill="currentColor"
      class="bx--select__arrow" width="16" height="16" viewBox="0 0 16 16"
      aria-hidden="true">
    <path d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"></path>
  </svg>
  {tmpl_icon_error}
</div>
"""
        else:
            template = """
<div class="bx--time-picker__select bx--select {class}">
  {tmpl_label}
  <select id="{id}" class="bx--select-input" {props}>
    {tmpl_items}
  </select>
  <svg focusable="false" preserveAspectRatio="xMidYMid meet"
      xmlns="http://www.w3.org/2000/svg" fill="currentColor"
      class="bx--select__arrow" width="16" height="16" viewBox="0 0 16 16"
      aria-hidden="true">
    <path d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"></path>
  </svg>
</div>
"""
        return self.format(template, values, context)


class TimezonePicker(BaseSelectFormNode):
    """Timezone Picker component

    To be used within TimePicker
    """
    def render_default(self, values, context):
        """Default html of the component.
        """
        if self.bound_field.errors:
            template = """
<div class="bx--time-picker__select bx--select bx--select--invalid {class}"
    data-invalid>
  {tmpl_label}
  <select id="{id}" class="bx--select-input" {props}>
    {tmpl_items}
  </select>
  <svg focusable="false" preserveAspectRatio="xMidYMid meet"
      xmlns="http://www.w3.org/2000/svg" fill="currentColor"
      class="bx--select__arrow" width="16" height="16" viewBox="0 0 16 16"
      aria-hidden="true">
    <path d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"></path>
  </svg>
  {tmpl_icon_error}
</div>
"""
        else:
            template = """
<div class="bx--time-picker__select bx--select {class}">
  {tmpl_label}
  <select id="{id}" class="bx--select-input" {props}>
    {tmpl_items}
  </select>
  <svg focusable="false" preserveAspectRatio="xMidYMid meet"
      xmlns="http://www.w3.org/2000/svg" fill="currentColor"
      class="bx--select__arrow" width="16" height="16" viewBox="0 0 16 16"
      aria-hidden="true">
    <path d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"></path>
  </svg>
</div>
"""
        return self.format(template, values, context)


class TimePicker(FormNode):
    """Time Picker component.

    The props ampm and tzinfo must be Django form fields
    """
    NODE_PROPS = ('light', 'ampm', 'tzinfo')
    "Extended Template Tag arguments."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        child_class = []
        child_props = {}

        if self.eval(self.kwargs.get('disabled'), context):
            values['label_class'].append('bx--label--disabled')
            child_props['disabled'] = True

        if self.eval(self.kwargs.get('light'), context):
            values['class'].append('bx--time-picker--light')
            child_class.append('bx--select--light')

        if child_class:
            child_props['class'] = ' '.join(child_class)

        # This is apparently needed
        self.nodelist.clear()

        ampm = self.eval(self.kwargs.get('ampm'), context)
        if ampm:
            self.nodelist.append(AmPmPicker(ampm, **child_props))

        tzinfo = self.eval(self.kwargs.get('tzinfo'), context)
        if tzinfo:
            self.nodelist.append(TimezonePicker(tzinfo, **child_props))


    def prepare_element_props(self, props, context):
        """Prepare html attributes for rendering the form element.
        """
        props['class'].append('bx--text-input')
        props['class'].append('bx--time-picker__input-field')
        props['pattern'] = r'(2[0-3]|1[0-9]|0?[0-9]):([1-5][0-9]|0?[0-9])(\\s)?'
        props['placeholder'] = 'hh:mm'
        props['maxlength'] = "5"

        if self.eval(self.kwargs.get('light'), context):
            props['class'].append('bx--text-input--light')


    def render_default(self, values, context):
        """Output html of the component.
        """
        light = self.eval(self.kwargs.get('light'), context)
        if self.bound_field.errors:
            if light:
                template = """
<div class="bx--form-item">
  <div class="bx--time-picker {class}" data-invalid>
    <div class="bx--time-picker__input">
      {tmpl_label}
      {tmpl_element}
    </div>
    {child}
  </div>
  <div class="bx--form-requirement">
    {tmpl_errors}
  </div>
  {tmpl_help}
</div>
"""
            else:
                template = """
<div class="bx--form-item">
  {tmpl_label}
  <div class="bx--time-picker {class}" data-invalid>
    <div class="bx--time-picker__input">{tmpl_element}</div>
    {child}
  </div>
  <div class="bx--form-requirement">
    {tmpl_errors}
  </div>
  {tmpl_help}
</div>
"""
        elif light and self.eval(self.kwargs.get('disabled'), context):
            template = """
<div class="bx--form-item">
  <div class="bx--time-picker {class}">
    <div class="bx--time-picker__input">
      {tmpl_label}
      {tmpl_element}
    </div>
    {child}
  </div>
  {tmpl_help}
</div>
"""
        else:
            template = """
<div class="bx--form-item">
  {tmpl_label}
  <div class="bx--time-picker {class}">
    <div class="bx--time-picker__input">{tmpl_element}</div>
    {child}
  </div>
  {tmpl_help}
</div>
"""
        return self.format(template, values, context)


    def render_tmpl_errors(self, values, context):
        """Dynamically render a part of the component's template.
        """
        tmpl_t = '<div class="bx--form-requirement__title">{title}.</div>'
        tmpl_s = '<p class="bx--form-requirement__supplement">{supplement}</p>'

        error_list = [self.bound_field.errors]
        ampm = self.eval(self.kwargs.get('ampm'), context)
        if ampm and ampm.errors:
            error_list.append(ampm.errors)
        tzinfo = self.eval(self.kwargs.get('tzinfo'), context)
        if tzinfo and tzinfo.errors:
            error_list.append(tzinfo.errors)

        items = []
        for error in chain(*error_list):
            title, supplement = error.split('.', 1)
            supplement = supplement.strip()
            items.append(tmpl_t.format(title=title))
            if supplement:
                items.append(tmpl_s.format(supplement=supplement.strip()))
        return '\n'.join(items)


components = {
    'TimePicker': TimePicker,
}

File: carbondesign/mithril-js/time_picker.js
import m from 'mithril/hyperscript';
//-
import { FormNode, render_errors } from './base';

class BaseSelectFormNode extends FormNode
{
    prepare(vnode, values, context)
    {
        values.props.push(['id', values.id]);
        values.props.push(['name', this.bound_field.name]);

        required = this.bound_field.field.required &&
                this.bound_field.form.use_required_attribute;
        if (required)
        {
            values.props.push(['required', '']);
        }
    }

    render_tmpl_icon_error(vnode, values, context)
    {
        return (
//##
m('svg.bx--select__invalid-icon',
  {
    focusable: false,
    preserveAspectRatio: 'xMidYMid meet',
    xmlns: 'http://www.w3.org/2000/svg',
    fill: 'currentColor',
    width: 16,
    height: 16,
    viewBox: '0 0 16 16',
    'aria-hidden': true,
  },
  [
    m('path',
      {
        d: 'M8,1C4.2,1,1,4.2,1,8s3.2,7,7,7s7-3.1,7-7S11.9,1,8,1z \
            M7.5,4h1v5h-1C7.5,9,7.5,4,7.5,4z M8,12.2 c-0.4,0-0.8-\
            0.4-0.8-0.8s0.3-0.8,0.8-0.8c0.4,0,0.8,0.4,0.8,0.8S8.4,\
            12.2,8,12.2z',
      }),
    m('path',
      {
        d: 'M7.5,4h1v5h-1C7.5,9,7.5,4,7.5,4z M8,12.2c-0.4,0-0.8-\
            0.4-0.8-0.8s0.3-0.8,0.8-0.8 c0.4,0,0.8,0.4,0.8,0.8S8.4,\
            12.2,8,12.2z',
        'data-icon-path': 'inner-path',
        opacity: 0,
      }),
  ])
//##
        );
    }

    render_tmpl_items(vnode, values, context)
    {
        let items = [];

        for (let [_, val, txt] of this.choices(vnode))
        {
            let props = {};
            if (this.boundValue && this.boundValue.indexOf(val) != -1)
            {
                props.selected = '';
            }
            items.push(
//##
m('option.bx--select-option', {value: val, ...props}, txt)
//##
            );
        }

        if (items.length)
        {
            return m.fragment(null, items);
        }
    }
}

class AmPmPicker extends BaseSelectFormNode
{
    render_default(vnode, values, context)
    {
        if (this.bound_field.errors)
        {
            return (
//##
m('div',
  {
    'class': `bx--time-picker__select bx--select bx--select--invalid ${values.class}`,
    'data-invalid': '',
  },
  [
    this.tmpl('label', ...arguments),
    m('select.bx--select-input', props, this.tmpl('items', ...arguments)),
    m('svg.bx--select__arrow',
      {
        focusable: false,
        preserveAspectRatio: 'xMidYMid meet',
        xmlns: 'http://www.w3.org/2000/svg',
        fill: 'currentColor',
        width: 16,
        height: 16,
        viewBox: '0 0 16 16',
        'aria-hidden': true,
      },
      m('path',
        {
          d: 'M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z',
        })),
    this.tmpl('icon_error', ...arguments),
  ])
//##
            );
        }

        return (
//##
m('div',
  {
    'class': `bx--time-picker__select bx--select ${values.class}`,
  },
  [
    this.tmpl('label', ...arguments),
    m('select.bx--select-input', props, this.tmpl('items', ...arguments)),
    m('svg.bx--select__arrow',
      {
        focusable: false,
        preserveAspectRatio: 'xMidYMid meet',
        xmlns: 'http://www.w3.org/2000/svg',
        fill: 'currentColor',
        width: 16,
        height: 16,
        viewBox: '0 0 16 16',
        'aria-hidden': true,
      },
      m('path',
        {
          d: 'M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z',
        })),
  ])
//##
        );
    }
}

class TimezonePicker extends BaseSelectFormNode
{
    render_default(vnode, values, context)
    {
        if (this.bound_field.errors)
        {
            return (
//##
m('div',
  {
    'class': `bx--time-picker__select bx--select bx--select--invalid ${values.class}`,
    'data-invalid': '',
  },
  [
    this.tmpl('label', ...arguments),
    m('select.bx--select-input', props, this.tmpl('items', ...arguments)),
    m('svg.bx--select__arrow',
      {
        focusable: false,
        preserveAspectRatio: 'xMidYMid meet',
        xmlns: 'http://www.w3.org/2000/svg',
        fill: 'currentColor',
        width: 16,
        height: 16,
        viewBox: '0 0 16 16',
        'aria-hidden': true,
      },
      m('path',
        {
          d: 'M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z',
        })),
    this.tmpl('icon_error', ...arguments),
  ])
//##
            );

        return (
//##
m('div',
  {
    'class': `bx--time-picker__select bx--select ${values.class}`,
  },
  [
    this.tmpl('label', ...arguments),
    m('select.bx--select-input', props, this.tmpl('items', ...arguments)),
    m('svg.bx--select__arrow',
      {
        focusable: false,
        preserveAspectRatio: 'xMidYMid meet',
        xmlns: 'http://www.w3.org/2000/svg',
        fill: 'currentColor',
        width: 16,
        height: 16,
        viewBox: '0 0 16 16',
        'aria-hidden': true,
      },
      m('path',
        {
          d: 'M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z',
        })),
  ])
//##
        );
    }
}

export class TimePicker extends FormNode
{
    NODE_PROPS = ['light', 'ampm', 'tzinfo']

    prepare(vnode, values, context)
    {
        let child_class = [];
        let child_props = {};

        if (vnode.attrs.disabled)
        {
            values.label_class.push('bx--label--disabled');
            values.props.push(['disabled', '']);
            child_props.disabled = '';
        }

        if (vnode.attrs.light)
        {
            values['class'].push('bx--time-picker--light');
            child_class.push('bx--select--light');
        }

        if (child_class.length)
        {
            child_props['class'] = child_class.join(' ');
        }

        this.nodelist = [];

        if (vnode.attrs.ampm)
        {
            this.nodelist.push(new AmPmPicker(Object.assign(
                    {field: vnode.attrs.ampm}, child_props)))
        }

        if (vnode.attrs.tzinfo)
        {
            this.nodelist.push(new TimezonePicker(Object.assign(
                    {field: vnode.attrs.tzinfo}, child_props)))
        }
    }

    prepare_element_props(vnode, props, default_props, context)
    {
        props['class'].push('bx--text-input');
        props['class'].push('bx--time-picker__input-field');
        props.pattern = '(2[0-3]|1[0-9]|0?[0-9]):([1-5][0-9]|0?[0-9])(\\\\s)?';
        props.placeholder = 'hh:mm';
        props.maxlength = '5'

        if (vnode.attrslight)
        {
            props['class'].push('bx--text-input--light');
        }
    }

    render_default(vnode, values, context)
    {
        let light = vnode.attrs.light;
        if (this.bound_field.errors)
        {
            if (light)
            {
                return (
//##
m('div.bx--form-item', null,
  [
    m('div',
      {
        'class': `bx--time-picker ${values.class}`,
        'data-invalid': '',
      },
      [
        m('div.bx--time-picker__input', null,
          [
            this.tmpl('label', ...arguments),
            this.tmpl('element', ...arguments),
          ]),
        values.child,
      ]),
    m('div.bx--form-requirement', null, this.tmpl('errors', ...arguments)),
    this.tmpl('help', ...arguments),
  ])
//##
                );
            }
            return (
//##
m('div.bx--form-item', null,
  [
    this.tmpl('label', ...arguments),
    m('div',
      {
        'class': `bx--time-picker ${values.class}`,
        'data-invalid': '',
      },
      [
        m('div.bx--time-picker__input', null,
            this.tmpl('element', ...arguments)),
        values.child,
      ]),
    m('div.bx--form-requirement', null, this.tmpl('errors', ...arguments)),
    this.tmpl('help', ...arguments),
  ])
//##
        }
        if (light && vnode.attrs.disabled)
        {
            return (
//##
m('div.bx--form-item', null,
  [
    m('div',
      {
        'class': `bx--time-picker ${values.class}`,
      },
      [
        m('div.bx--time-picker__input', null,
          [
            this.tmpl('label', ...arguments),
            this.tmpl('element', ...arguments),
          ]),
        values.child,
      ]),
    this.tmpl('help', ...arguments),
  ])
//##
            );
        }
        return (
//##
m('div.bx--form-item', null,
  [
    this.tmpl('label', ...arguments),
    m('div',
      {
        'class': `bx--time-picker ${values.class}`,
      },
      [
        m('div.bx--time-picker__input', null,
            this.tmpl('element', ...arguments)),
        values.child,
      ]),
    this.tmpl('help', ...arguments),
  ])
//##
        );
    }

    render_tmpl_errors(vnode, values, context)
    {
        let bound_fields = [this.bound_field];
        if (vnode.attrs.ampm)
        {
            bound_fields.push(vnode.attrs.ampm);
        }
        if (vnode.attrs.tzinfo)
        {
            bound_fields.push(vnode.attrs.tzinfo);
        }
        return render_errors(bound_fields);
    }
}
